src/components/Tables/Table.tsx

```diff
@@ -0,0 +1,149 @@
+"use client";
+
+import React, { useState, useEffect } from "react";
+import DataTable, { TableColumn } from "react-data-table-component";
+import TableFilterBar, { FilterText } from "@/components/Tables/TableFilterBar";
+
+interface Row {
+    checkBoxKey: number;
+    [key: string]: string | number;
+}
+interface Headers {
+    [key: string]: string;
+}
+
+export interface Datum {
+    [key: string]: string | number | null;
+}
+
+interface Props {
+    data: Datum[];
+    headers: Headers;
+}
+
+const itemIncludesFilterText = (headers: Headers, item: Datum, filterText: FilterText): boolean => {
+    for (const key of Object.keys(headers)) {
+        if (
+            !(item[key] ?? "")
+                .toString()
+                .toLowerCase()
+                .includes((filterText[key] ?? "").toLowerCase())
+        ) {
+            return false;
+        }
+    }
+    return true;
+};
+
+const filteredNoNullItems = (headers: Headers, data: Datum[], filterText: FilterText): Row[] => {
+    let row = 0;
+    return data.reduce((filteredNoNullList: Row[], item: Datum, currentIndex: number) => {
+        if (itemIncludesFilterText(headers, item, filterText)) {
+            const noNullItem: Row = { checkBoxKey: currentIndex };
+            for (const key of Object.keys(headers)) {
+                noNullItem[key] = item[key] ?? "";
+            }
+            noNullItem.id = row;
+            filteredNoNullList.push(noNullItem);
+            row++;
+        }
+        return filteredNoNullList;
+    }, []);
+};
+
+const Table: React.FC<Props> = (props) => {
+    const [filterText, setFilterText] = useState<FilterText>({});
+
+    const [selectCheckBox, setSelectCheckBox] = useState(
+        new Array<boolean>(props.data.length).fill(false)
+    );
+    const [selectAllCheckBox, setSelectAllCheckBox] = useState(false);
+
+    const toggleAllCheckBox = (): void => {
+        setSelectCheckBox(new Array<boolean>(props.data.length).fill(!selectAllCheckBox));
```

Comment:
does this work with pagination? (can you make a test for this? you should test the checkbox stuff _really thoroughly_ because it's something we've made, rather than just trusting that the library we `npm install`ed worked.

Reply:
Made